d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingState.java

```diff
@@ -0,0 +1,133 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.subsetting;
+
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.http.annotation.GuardedBy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class SubsettingState
+{
+  private static final Logger LOG = LoggerFactory.getLogger(SubsettingState.class);
+  private final Object _lock = new Object();
+  private final SubsettingStrategyFactory _subsettingStrategyFactory;
+
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Map<URI, TrackerClient>>> _weightedSubsetsCache;
+
+  @GuardedBy("_lock")
+  private long _version;
+
+  @GuardedBy("_lock")
+  private long _peerClusterVersion;
+
+  @GuardedBy("_lock")
+  private long _minClusterSubsetSize;
```

Comment:
Does one subsettingState maps to one service in the cluster or the whole peer cluster? Based on the implementation, it seems that there is only one subsettingState for the whole peer cluster, but technically the minClusterSubsetSize can be different in different d2Services?

Reply:
Good point! Fixed by adding a service -> minClusterSubsetSize map